Add a solution for oop task

Sofiia-Viktoriia opened the pull request.

Sofiia-Viktoriia committed:
Add a solution for oop task

V1ctr0 commented on PracticeProject/Employee.cs:
```diff
@@ -0,0 +1,45 @@
+    {
+        private string _name;
+        private double _salary;
+        protected double _bonus;
```
Повинно бути приватним

Sofiia-Viktoriia replied:
Done

V1ctr0 commented on PracticeProject/SalesPerson.cs:
```diff
@@ -0,0 +1,25 @@
+            else if (_percent > 100)
+            {
+                _bonus *= 2;
+            }
```
`SetBonus `і є по факту твій сетер, тож використовуй його. На додачу, як пофіксиш, використай `SetBonus` один раз

Sofiia-Viktoriia replied:
Done

V1ctr0 commented on PracticeProject/Manager.cs:
```diff
@@ -0,0 +1,25 @@
+            else if (_quantity > 100)
+            {
+                _bonus += 500;
+            }
```
Теж саме, що і для цього методу в `SalesPerson `класі

Sofiia-Viktoriia replied:
Done

V1ctr0 commented on PracticeProject/Company.cs:
```diff
@@ -0,0 +1,38 @@
+            {
+                sum += employee.ToPay();
+            }
+            return sum;
```
Валідний варіант, але спробуй виконати цю ж дію в один рядок з використанням методу із `System.Linq`

Sofiia-Viktoriia replied:
Done

Sofiia-Viktoriia committed:
Make _bonus field private
Update SetBonus method
Replace foreach cycle with linq expression in TotalToPay method

V1ctr0 commented on PracticeProject/Company.cs:
```diff
@@ -0,0 +1,37 @@
+            double sum = (
+                from employee in _employees
+                select employee.ToPay()
+            ).Sum();
```
В більшості випадків особисто я зтикаюся з варіантом запису через методи, а не запити (однотонність з іншим С# кодом), хоча варіант через запити може використовуватися при роботі з БД, або коли потрібні громіздкі операції з джоінами данних.

Теж саме можна зробити значно простіше `_employees.Sum(x => x.ToPay())`

Sofiia-Viktoriia replied:
Done

Sofiia-Viktoriia committed:
Refactor TotalToPay method
